### PR TITLE
refactor command and error handling

### PR DESCRIPTION
**Pull request by Yakiyo**
tldr my changes are
- refactor commands to separate folder
- move command handling to the `cli.rs` rather than main
- main func only runs the command handler and handles any error. The main command handling is with the cli struct
- pretty print pkg name, version, path to project and script before the script is ran. 

The current implementations all fused into the main and cli file. Moving them to a separate folder makes it more readable, a separate file for each command makes it easier for anyone to figure out whats going on where. I left the main logic untouched. But added the `name` and `version` property to node manifest. pn now prints the `pkg@version script /path/to/project` thingy like pnpm does before running a script. Global cli flags like the `--workspace-root` were moved to a common `Config` struct which is flattened by clap. This allows us to pass around the Config struct to subcommand handlers.

Errors could use something like [`anyhow`](https://github.com/dtolnay/anyhow) or [`thiserror`](https://github.com/dtolnay/thiserror) instead of implement display on enum variants. thiserror seems better than using anyhow since its returning error types and parsing it rather than just strings. I'd be happy to implement that in case its acceptable. It could also potentially ease solving #16

### CODE REVIEW

**Comment by KSXGitHub** [comment]
> Errors could use something like [`anyhow`](https://github.com/dtolnay/anyhow) or [`thiserror`](https://github.com/dtolnay/thiserror) instead of implement display on enum variants.

I don't like `thiserror`. See this comment: https://github.com/pnpm/pn/pull/24#issue-1751297851

Also, whilst I could have used `derive_more::Error` to derive the `Error` trait, I chose not to. Because `pn` is a binary, not a library. And `pn` doesn't use extra information provided by the `Error` trait, so I only derive `Display` to print it.

---

Regarding the modules, I personally prefer this structure over `mod.rs`:

```
.
├── lib.rs
├── my_module
│   ├── sub1.rs
│   └── sub2.rs
└── my_module.rs
```

There are 2 reasons:
* Seeing a tab named `mod.rs` in the editor doesn't tell me about which module it belongs to.
* <kbd>Ctrl-P</kbd> doesn't work well if there are many `mod.rs`.

I recognize that other projects use `mod.rs`. But I believe this is simply because they were started when Rust didn't yet support my preferred structure. Either that, or they have not recognize the superiority of my way.

Anyway, @zkochan, this is your project, you should make this choice: Do you prefer `mod.rs` or not-`mod.rs`?

For context, `mod.rs` is like `index.js` in Node.js so the structure would be like this:

```
.
├── lib.rs
└── my_module
    ├── mod.rs
    ├── sub1.rs
    └── sub2.rs
```

_(if I start another Node.js project today, I would not create `index.js` either, save for one at the top-level)_

---

> But added the `name` and `version` property to node manifest. pn now prints the `pkg@version script /path/to/project` thingy like pnpm does before running a script.

I like feature commits. Can you this commit to a separate PR without all the refactoring? I sense that the refactor discussion is going to be lengthy.

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
https://github.com/pnpm/pn/pull/39#issuecomment-1595763526

File: src/commands/run.rs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+#[serde(rename_all = "kebab-case")]
+pub struct NodeManifest {
+    #[serde(default)]
+    name: String,
+
+    #[serde(default)]
+    scripts: HashMap<String, String>,
+
+    #[serde(default)]
+    version: String,
```
Comment (KSXGitHub) [inline src/commands/run.rs:65]: Both `pnpm run` and `npm run` functions without `name` and `version` in `package.json`. There is no reason to make them mandatory field.

File: src/commands/run.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            r#"{"name": "", "scripts": {"test": "echo hello world"}, "version": ""}"#,
+        )
+        .unwrap();
+
+        let package_manifest = read_package_manifest(&package_json_path).unwrap();
+
+        let received = serde_json::to_string_pretty(&package_manifest).unwrap();
+        let expected = serde_json::json!({
+            "name": "",
+            "version": "",
```
Comment (KSXGitHub) [inline src/commands/run.rs:140]: If `name` and `version` would become optional, this can be removed.

```suggestion
```

File: src/commands/run.rs (hunk 1 above)
Comment (KSXGitHub) [reply src/commands/run.rs:65]: Is there any reason `name` and `version` are not next to each other?

**Comment by zkochan** [comment]
> Anyway, @zkochan, this is your project, you should make this choice: Do you prefer mod.rs or not-mod.rs?

so without `mod.rs`, how would I create a module named `my_module`? By creating `my_module.rs`? So there would be both a directory called `my_module` and an rs file name the same way?

**Comment by KSXGitHub** [comment]
> By creating `my_module.rs`? So there would be both a directory called `my_module` and an rs file name the same way?

Yes. There would be a file named `my_module.rs` and a directory named `my_module` (without `.rs`) that contains all submodules of `my_module`. Reference: [parallel-disk-usage/src](https://github.com/KSXGitHub/parallel-disk-usage/tree/1a8ad8dd068d813341cc07b56ec1b9f795cacdb4/src).

**Comment by KSXGitHub** [comment]
Also, if `my_module` doesn't have any submodule, the directory would not be created.

> how would I create a module named `my_module`?

I create them using the "code actions" in VS Code + rust-analyzer.

**Comment by zkochan** [comment]
ok, let's do it your way. With `my_module.rs`

File: src/cli.rs [hunk 3]
```diff
@@ -1,49 +1,139 @@
...
+    pub config: Config,
+
     /// Command to execute.
     #[clap(subcommand)]
     pub command: Command,
 }
 
+/// Global configurations to be passed down to subcommands
+#[derive(Parser, Debug)]
+pub struct Config {
```
Comment (KSXGitHub) [inline src/cli.rs:21]: Rename this to `GlobalOptions`.

File: src/commands/run.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+#[clap(rename_all = "kebab-case")]
+pub struct Run {
+    /// Name of the package script to run.
+    pub script: String, // Not OsString because it would be compared against package.json#scripts
+
+    /// Arguments to pass to the package script.
+    pub args: Vec<OsString>,
+}
+
+impl CommandTrait for Run {
```
Comment (KSXGitHub) [inline src/commands/run.rs:27]: It is Rust's convention that a trait's name should match it's sole method name (for example: `Into::into`, `AsRef::as_ref`, `ToOwned::to_owned`, etc).

Therefore, I suggest the trait be renamed `ExecuteCommand` and the method `execute_command`.

File: src/cli.rs [hunk 5]
```diff
@@ -1,49 +1,139 @@
...
 }
 
-#[derive(Debug, Args)]
-#[clap(rename_all = "kebab-case")]
-pub struct PassedThroughArgs {
-    pub args: Vec<OsString>,
+impl Cli {
+    pub fn run(self) -> Result<(), MainError> {
+        match self.command {
+            Command::Run(e) => e.run(self.config),
```
Comment (KSXGitHub) [inline src/cli.rs:52]: So you defined a trait only to use it once? I suggest you remove the trait.

File: src/commands/run.rs (hunk 4 above)
Comment (KSXGitHub) [reply src/commands/run.rs:27]: It turns out that this trait was a waste after all: https://github.com/pnpm/pn/pull/39/files#r1233114060

File: src/main.rs [hunk 6]
```diff
@@ -1,249 +1,26 @@
-use ansi_term::Color::{Black, Red};
+use ansi_term::Colour::{Black, Red};
```
Comment (KSXGitHub) [inline src/main.rs:1]: We have switched to `yansi` recently.

**Comment by KSXGitHub** [comment]
After examining your code. It turns out that many of your refactoring only makes the code more verbose than necessary.

I don't want you to waste your effort unnecessarily, so I have a suggestion: If you plan to refactor code in the future, please open an issue to discuss with project members first to see if a refactor is necessary and what structure/paradigm should the new code adhere to.

I do like that you add the feature regarding `name` and `version`, however. I'd greatly appreciate it if you could create a PR with this feature and none of the refactoring.

File: src/commands/run.rs (hunk 1 above)
Comment (Yakiyo) [reply src/commands/run.rs:65]: there actually is. initally i did put them side by side but the read_package_manifest test was failing. the json! macro ends up sorting the `expected` value alphabetically while parsing the nodemanifest struct to json sorts it in the same way as the struct is defined. so the test ends up failing. im unusure how else it could be changed. we could maybe only compare the scripts hashmap with another hashmap or something.

File: src/cli.rs (hunk 5 above)
Comment (Yakiyo) [reply src/cli.rs:52]: i added the trait cz i thought thered be more commands added later on or something. ig for now it seems useless. so ill remove it then.

File: src/commands/run.rs (hunk 2 above)
Comment (Yakiyo) [reply src/commands/run.rs:140]: they are optional but it does end up failing the test since converting the NodeManifest struct to json includes the empty name and version in the output.

File: src/commands/run.rs (hunk 1 above)
Comment (Yakiyo) [reply src/commands/run.rs:65]: Heres an image of the test. the first one is the `received` json and the second is `expected`
image

thus it failed
image

File: src/cli.rs (hunk 3 above)
Comment (Yakiyo) [reply src/cli.rs]: Done. renamed the struct to GlobalOptions and the Cli prop to global_options

File: src/commands/run.rs (hunk 4 above)
Comment (Yakiyo) [reply src/commands/run.rs]: Removed the trait entirely [here](https://github.com/pnpm/pn/pull/39/commits/576f80f65a8ce78bc6eaf2ab5cf3c4604e033d92)

File: src/main.rs (hunk 6 above)
Comment (Yakiyo) [reply src/main.rs]: My bad i did see the pr for yansi but forgot about it. Removed all ansi_term and used yansi instances

**Comment by Yakiyo** [comment]
> After examining your code. It turns out that many of your refactoring only makes the code more verbose than necessary.
> 
> I don't want you to waste your effort unnecessarily, so I have a suggestion: If you plan to refactor code in the future, please open an issue to discuss with project members first to see if a refactor is necessary and what structure/paradigm should the new code adhere to.
> 
> I do like that you add the feature regarding `name` and `version`, however. I'd greatly appreciate it if you could create a PR with this feature and none of the refactoring.

Makes sense. Then do i just close this pr and open a new one with just the name/version change?

**Comment by Yakiyo** [comment]
The refactors here seems to be unnecessary. I dont mind, i should have opened an issue first, my bad. According to @KSXGitHub's suggestion i opened a pr with only the script name-version change #43 without any of the refactorings. Unless theres anything left, this pr should be closed ig.

File: src/commands/run.rs (hunk 2 above)
Comment (KSXGitHub) [reply src/commands/run.rs:139]: You can compare `serde_json::Value`: Replace `to_string_pretty` with `to_value` in the `received` expression and remove `to_string_pretty` in the `expected` expression.

Comment (KSXGitHub) [reply src/commands/run.rs:139]:
```rust
        let received = serde_json::to_value(&package_manifest).unwrap();
        let expected = serde_json::json!({
            "name": "",
            "version": "",
            "scripts": {
                "test": "echo hello world"
            }
        });
```
